### ISSUE TITLE
`graphene::Rect::contains_point(&self, p: &Point)` is wrong for Points above the Rect [BUG]

### ISSUE THREAD

**Issue by flxzt** [gtk-rs/gtk-rs-core#199]
**Bug description**
`Rect::contains_point(&self, p: &Point)` returns `true` for points which are above the rectangle. It should return `false` when they are not inside it.
This is the code that can reproduce the issue:

```
use gtk4::graphene::{Point, Rect};

fn main() {
    let rect = Rect::new(100., 100., 100., 100.);

    let right = Point::new(250., 150.);
    let below = Point::new(150., 50.);
    let left = Point::new(50., 150.);
    let above = Point::new(150., 250.);

    assert!(!rect.contains_point(&right));
    assert!(!rect.contains_point(&below));
    assert!(!rect.contains_point(&left));
    assert!(!rect.contains_point(&above));
}
```

**Backtrace**

```
thread 'main' panicked at 'assertion failed: !rect.contains_point(&above)', src/main.rs:14:5
stack backtrace:
   0:     0x55897d544380 - std::backtrace_rs::backtrace::libunwind::trace::ha5edb8ba5c6b7a6c
                               at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/../../backtrace/src/backtrace/libunwind.rs:90:5
   1:     0x55897d544380 - std::backtrace_rs::backtrace::trace_unsynchronized::h0de86d320a827db2
                               at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x55897d544380 - std::sys_common::backtrace::_print_fmt::h97b9ad6f0a1380ff
                               at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/sys_common/backtrace.rs:67:5
   3:     0x55897d544380 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h14be7eb08f97fe80
                               at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/sys_common/backtrace.rs:46:22
   4:     0x55897d55e2cf - core::fmt::write::h2ca8877d3e0e52de
                               at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/fmt/mod.rs:1094:17
   5:     0x55897d542e85 - std::io::Write::write_fmt::h64f5987220b618f4
                               at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/io/mod.rs:1584:15
   6:     0x55897d545fcb - std::sys_common::backtrace::_print::h7f1a4097308f2e0a
                               at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/sys_common/backtrace.rs:49:5
   7:     0x55897d545fcb - std::sys_common::backtrace::print::h1f799fc2ca7f5035
                               at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/sys_common/backtrace.rs:36:9
   8:     0x55897d545fcb - std::panicking::default_hook::{{closure}}::hf38436e8a3ce1071
                               at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/panicking.rs:208:50
   9:     0x55897d545a9d - std::panicking::default_hook::he2f8f3fae11ed1dd
                               at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/panicking.rs:225:9
  10:     0x55897d5465dd - std::panicking::rust_panic_with_hook::h79a18548bd90c7d4
                               at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/panicking.rs:591:17
  11:     0x55897d546147 - std::panicking::begin_panic_handler::{{closure}}::h212a72cc08e25126
                               at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/panicking.rs:495:13
  12:     0x55897d54481c - std::sys_common::backtrace::__rust_end_short_backtrace::hbd6897dd42bc0fcd
                               at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/sys_common/backtrace.rs:141:18
  13:     0x55897d5460d9 - rust_begin_unwind
                               at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/panicking.rs:493:5
  14:     0x55897d534d91 - core::panicking::panic_fmt::h77ecd04e9b1dd84d
                               at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/panicking.rs:92:14
  15:     0x55897d534cdd - core::panicking::panic::h60569d8a39169222
                               at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/panicking.rs:50:5
  16:     0x55897d5358f6 - graphenetest::main::hd472e6662594c6f2
                               at /run/media/felixz/Daten/source/rust/graphenetest/src/main.rs:14:5
  17:     0x55897d5355db - core::ops::function::FnOnce::call_once::h979fe99887ecba2d
                               at /home/felixz/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227:5
  18:     0x55897d535a3e - std::sys_common::backtrace::__rust_begin_short_backtrace::h0e90eea8e5c5bd6e
                               at /home/felixz/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:125:18
  19:     0x55897d535641 - std::rt::lang_start::{{closure}}::hd476fe4938731590
                               at /home/felixz/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:49:18
  20:     0x55897d5469da - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::hc4354216bf39217c
                               at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/ops/function.rs:259:13
  21:     0x55897d5469da - std::panicking::try::do_call::hb68eb312780385cf
                               at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/panicking.rs:379:40
  22:     0x55897d5469da - std::panicking::try::h22b8e08595060b8b
                               at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/panicking.rs:343:19
  23:     0x55897d5469da - std::panic::catch_unwind::hc64f1a6a0e71b1fc
                               at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/panic.rs:431:14
  24:     0x55897d5469da - std::rt::lang_start_internal::h4461fc58637f04f8
                               at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/rt.rs:34:21
  25:     0x55897d535620 - std::rt::lang_start::h2c890320d1935a21
                               at /home/felixz/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:48:5
  26:     0x55897d535a2c - main
  27:     0x7f92562d8b75 - <unknown>
  28:     0x55897d53548e - _start
  29:                0x0 - <unknown>
```

**Comment by sdroege** [comment]
This would be a bug in graphene, the underlying C library. Please report it at https://github.com/ebassi/graphene

Thanks!

**Comment by felixzwettler** [comment]
Tests for the C library are passing (look [here](https://github.com/ebassi/graphene/issues/234) ). Emmanuel said the cause must be somewhere in the bindings. So here I am again. :)

**Comment by sdroege** [comment]
Curious! That needs some printfs or checking in a debugger as a first step to see what exactly arrives in the C code. It's apparently not what is created on the Rust side.

**Comment by sdroege** [comment]
Something's very strange here:

```
(gdb) list
560	  graphene_rect_t rr;
561	
562	  rr = *r;
563	  graphene_rect_normalize_in_place (&rr);
564	
565	  return p->x >= rr.origin.x &&
566	         p->y >= rr.origin.y &&
567	         p->x <= (rr.origin.x + rr.size.width) &&
568	         p->y <= (rr.origin.y + rr.size.height);
569	}
(gdb) print *p
$11 = {x = 150, y = 250}
(gdb) print rr
$8 = {origin = {x = 100, y = 100}, size = {width = 100, height = 100}}
(gdb) print p->x >= rr.origin.x
$12 = 1
(gdb) print p->y >= rr.origin.y
$13 = 1
(gdb) print p->x <= (rr.origin.x + rr.size.width)
$14 = 1
(gdb) print p->y <= (rr.origin.y + rr.size.height)
$15 = 0
```

So the overall return value should be 0... but for some reason it ends up being true. I'll debug more tomorrow.

@bilelmoussaoui we should also add proper `Debug` impls for all the graphene types :)

**Comment by bilelmoussaoui** [comment]
For that we need https://github.com/gtk-rs/gir/issues/1202 first :/

**Comment by sdroege** [comment]
For non-GObjects that is possible already:

```toml
[[object]]
name = "Gst.Format"
status = "generate"
    [[object.derive]]
    name = "Debug, PartialEq, Eq, PartialOrd, Ord, Hash"
```

You give the whole list you want to have derived there.

**Comment by sdroege** [comment]
> So the overall return value should be 0... but for some reason it ends up being true (some random non-0 number and not 1!). I'll debug more tomorrow.

It works fine with a debug build of graphene, oh no.

**Comment by sdroege** [comment]
Printing the return value before returning inside `graphene_rect_contains_point()` also makes it not happen.

**Comment by sdroege** [comment]
The problem is that graphene returns `bool` (1 byte) and not `gboolean` (an `int`). Somehow we need to detect that from the .gir file and use `bool` instead of `glib_sys::gboolean` for the return values in the -sys crate, and stop using `from_glib`.

This seems all a bit suboptimal.

**Comment by sdroege** [comment]
Wrote a comment in the graphene issue, let's see how this can be solved. It's surprising that nobody else ran into this yet, maybe means that nobody really uses the graphene bindings.

**Comment by sdroege** [comment]
This is fixed in 0.14.8 now.
